go/libraries/doltcore/doltdb/global_table_editor.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doltdb
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/schema"
+)
+
+// GetRootFunc returns the root value that the GlobalTableEditor should be using.
+type GetRootFunc func(ctx context.Context) (*RootValue, error)
+
+// GlobalTableEditor handles all edit operations on a table that may also update other tables. Serves as coordination
+// for LocalTableEditors.
+type GlobalTableEditorImpl struct {
+	DisableForeignKeyChecks bool // If disabled, then all foreign key checks and updates are ignored
+
+	getRoot    GetRootFunc
+	tables     map[string]*LocalTableEditor
+	writeMutex *sync.RWMutex // This mutex is specifically for changes that affect the GlobalTE or all LocalTEs
+}
+
+var globalTableEditor = &GlobalTableEditorImpl{}
+
+func init() {
+	globalTableEditor.tables = make(map[string]*LocalTableEditor)
+	globalTableEditor.writeMutex = &sync.RWMutex{}
+}
+
+// GlobalTableEditor returns the global table editor.
+func GlobalTableEditor() *GlobalTableEditorImpl {
```

Comment:
This is the wrong abstraction to use here. Rather than a global variable shared by SQL all sessions, this needs to be local to a particular session or transaction. Suggest TableEditSession for a start (it's not really a Transaction yet, so probably don't call it that). Instead of accessing this globally, instantiate one for each DoltSession to make available to SQL methods. Could do this lazily if you feel ambitious.